credential/concept/template.go

```diff
@@ -0,0 +1,319 @@
+				if err := ct.processValue(sv, nextPath); err != nil {
+					return err
+				}
+			} else {
+				gjs := gojsonq.New().FromInterface(v)
+				if err := ct.mapR(gjs, nextPath); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (ct *template) processValue(value string, currentPath string) error {
+	processValue := templateString(value)
+
+	if !processValue.isValid() {
+		return fmt.Errorf("values has incorrect format at %s: %s", currentPath, value)
+	}
+
+	if processValue.isHardCoded() {
+		ct.fixedValues[currentPath] = processValue.String()
+	} else {
+		ct.conceptIndexMapping[processValue.String()] = currentPath
+		ct.reverseIndexMapping[currentPath] = processValue.String()
+	}
+
+	if processValue.hasIndex() {
+		for _, index := range processValue.indices() {
+			ct.addIndex(index[0], index[1], currentPath)
+		}
+	}
+
+	return nil
+}
+
+// addIndex expects lvl Indices to start at 1
+func (ct *template) addIndex(lvl1 int, lvl2 int, jsonPath string) {
+	//always replace & copy => less if/else logic
+	lvl1Size := max(len(ct.indices), lvl1)
+	newLvl1 := make([][]string, lvl1Size)
+	for i, v := range ct.indices {
+		newLvl1[i] = v
+	}
+
+	lvl2Size := max(len(newLvl1[lvl1-1]), lvl2)
+	newLvl2 := make([]string, lvl2Size)
+	for i, v := range newLvl1[lvl1-1] {
+		newLvl2[i] = v
+	}
+	ct.indices = newLvl1
+	ct.indices[lvl1-1] = newLvl2
+	ct.indices[lvl1-1][lvl2-1] = ct.reverseIndexMapping[jsonPath]
+}
+
+// transform a VC to a concept given the template mapping
+func (ct *template) transform(VC did.VerifiableCredential) (Concept, error) {
+	// remove type as this will be hardcoded later
+	VC.Type = nil
+
+	vcJSON, err := json.Marshal(VC)
+	if err != nil {
+		return nil, err
+	}
+
+	// json parsing
+	jq := gojsonq.New()
+	jq.FromString(string(vcJSON))
+
+	// result target
+	c := Concept{
+		TypeField: ct.fixedValues[TypeField],
+	}
+	err = ct.transR(jq, "", c)
+
+	return c, err
+}
+
+func (ct *template) transR(jq *gojsonq.JSONQ, currentPath string, c Concept) error {
+	val := jq.Get()
+
+	if jq.Error() != nil {
+		return jq.Error()
+	}
+
+	// we do not support mapped values within lists
+	if _, ok := val.([]interface{}); ok {
+		return errors.New("json arrays are not supported")
+	}
+
+	if m, ok := val.(map[string]interface{}); ok {
+		for k, v := range m {
+			nextPath := path(currentPath, k)
+
+			if v == nil {
+				continue
+			}
+
+			if sv, ok := v.(string); ok {
+				if m, ok := ct.reverseIndexMapping[nextPath]; ok {
+					c.SetValue(m, sv)
+				}
+			} else {
+				gjs := gojsonq.New().FromInterface(v)
+				if err := ct.transR(gjs, nextPath, c); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (cs templateString) isHardCoded() bool {
+	return !strings.Contains(string(cs), "<<")
+}
+
+func (cs templateString) hasIndex() bool {
+	return strings.Contains(string(cs), "@")
+}
+
+var regex, _ = regexp.Compile("((<<[a-zA-Z\\.]+>>)|([a-zA-Z\\.]+))(@({[1-9](_[1-9])?})(,{[1-9](_[1-9])?})*)?")
```

Comment:
give the var a more descriptive name than just `regex`, which almost collides with the package name (what does it match?)